[Feature/register] 회원 등록 기능 구현

yyysolhhh opened the pull request:
### Summary
회원 등록 기능 구현

### Description
<!-- 구체적인 작업 내용, 필요시 이미지 첨부 -->
- 프로젝트 구조 변경
- 공통 응답 dto 정의
- GlobalExceptionHandler, CustomErrorCode 생성하여 예외 응답 처리
- UserRegisterResponse DTO 생성
- User 등록하는 API 구현
- 이메일 중복 검사
- jwt sequence diagram 추가
- AdminInitializerTest, UserRepositoryTest 추가


### Issue Number or Link
<!-- 예시: closes #issue-number -->
closes #22 #23 #24

yyysolhhh committed:
docs: jwt sequence diagram mermaid로 작성

- 각 상황별 sequence diagram 작성

yyysolhhh committed:
chore: 프로젝트 구조 변경

- domain - entity, repository
- interfaces - controller, dto

yyysolhhh committed:
feat: CSRF, 기본 formLogin 비활성화

SecurityConfig
- csrf, formLogin disable

yyysolhhh committed:
feat: 공통 api response dto 생성

Api
- code, status, body
- ok, create 메서드 정의

yyysolhhh committed:
chore: User, Role 경로 수정

- import 경로 수정됨

yyysolhhh committed:
feat: UserRegisterResponse DTO 생성

UserRegisterResponse
- 사용자 등록 후 응답 DTO
- id 반환

yyysolhhh committed:
feat: User 등록 api 구현

UserService
- registerUser 메서드에서 User 등록

UserController
- registerUser 메서드 PostMapping
- /api/user/register

yyysolhhh committed:
feat: CustomErrorCode에 EMAIL_ALREADY_EXISTS 추가

- 409 conflict
- Email already exists
- 이메일 중복시 발생하는 에러 코드 정의

yyysolhhh committed:
feat: 커스텀 예외처리 구현

CustomException
- CustomErrorCode에 있는 에러 발생

GlobalExceptionHandler
- 커스텀 에러 발생시 에러 응답 반환

yyysolhhh committed:
feat: Api 공통 응답에 error 메서드 구현

- 에러 발생시 error 메서드로 응답 반환

yyysolhhh committed:
chore: import 경로 수정

- UserService 경로 변경에 따라 import 경로 수정

yyysolhhh committed:
feat: 중복 이메일 유효성 검사 기능 구현

UserService
- validateEmailDuplicate 메서드 구현
- 중복 이메일이 있을 경우 에러 발생

UserRepository
- existsByEmail 메서드 등록

yyysolhhh committed:
chore: application.yml 테스트 환경 설정 파일 추가

build.gradle
- h2 의존성 추가

application-test.yml
- datasource, logging 설정 추가

yyysolhhh committed:
chore: 마지막에 빈줄 추가

yyysolhhh committed:
test: UserRepositoryTest, AdminInitializerTest 작성

- existsByRole true, false 테스트
- existsByEmail true, false 테스트
- 관리자 계정 없을때 있을때 테스트

yyysolhhh committed:
chore: User 객체 생성하는 메서드 이름 of로 변경

- 정적팩토리메서드 명명 규칙

yyysolhhh committed:
chore: 파일 마지막줄에 빈줄 추가

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+
+    @ExceptionHandler(value = {CustomException.class})
+    protected Api<?> handleCustomException(CustomException e) {
+        log.error(e.getMessage());
```
ExceptionHandler가 캐치하면 로깅이 자동으로 되는것으로 알고있습니다.. 같이 확인한번 해봐요!

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/config/SecurityConfig.java:
```diff
@@ -2,12 +2,26 @@
 
+    @Bean
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        http
```
이 부분은 의도하고 사용하신 부분인가요? 포맷팅이 부자연스러워 보입니다.
http.csrf(AbstractHttpConfigurer::disable)  
&nbsp; &nbsp;.formLogin(AbstractHttpConfigurer::disable)
이런식으로 작성하면 어떨까요?

yyysolhhh replied:
수정했습니다.

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/common/Api.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public static <T> Api<T> create(@Nullable final T body) {
+        HttpStatus created = HttpStatus.CREATED;
```
http status와 메시지는 이미 Header에 전송되니, 해당 API 공통 응답 모델은 없애는게 어떨까요?
만든다면 페이지네이션 같은 부분을 공통으로 만드는 것이 적합해보입니다.

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+    @ExceptionHandler(value = {CustomException.class})
+    protected Api<?> handleCustomException(CustomException e) {
+        log.error(e.getMessage());
+        return Api.error(e);
```
이렇게 되면 항상 HTTP Status는 200 OK고,
응답 body에 code와 Message가 담기게 됩니다. 
저는 이것이 Restful하다고 생각되지 않습니다.

`
return ResponseEntity.status(httpStatus).body(e..);  
`

위와같은 방식은 어떨까요? Restful하게 status와 메시지를 확인할 수 있을 것 같습니다.

yyysolhhh replied:
ResponseEntity를 반환하는 것으로 수정했습니다.

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/user/domain/entity/User.java:
```diff
@@ -8,11 +8,13 @@
 import lombok.NoArgsConstructor;
 
 @Getter
+@Table(name = "\"user\"")
```
"\"로 감싸는 이유가 있나요?

yyysolhhh replied:
sql select 쿼리에서 문제가 생겨서 추후 해결하고 수정하겠습니다.

yyysolhhh committed:
feat: 에러 응답 ResponstEntity로 감싸서 응답처리

GlocalExceptionHandler
- ResponseEntity로 응답 반환

ExceptionResponse
- json 형식으로 응답하기 위해 DTO 생성

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/user/domain/service/UserService.java:
```diff
@@ -0,0 +1,50 @@
+    @Transactional
+    public Long registerUser(UserRegisterRequest request) {
+        validateEmailDuplicate(request.email());
+        User user = User.of(
```
userRepository.save(user).getId();
or
savedUser = userRepository.save(user);
savedUser.getId()
위처럼 db에 저장된 이후 생긴 id 값을 반환해야 할 것 같습니다.

yyysolhhh replied:
일단 수정했는데 저 방법도 가능하긴 해서 어떤게 더 좋은 방법인지 고민이 필요합니다.

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/user/interfaces/controller/UserController.java:
```diff
@@ -0,0 +1,27 @@
+            @RequestBody UserRegisterRequest userRegisterDto
+    ) {
+        Long userId = userService.registerUser(userRegisterDto);
+        return Api.create(UserRegisterResponse.of(userId));
```
마찬가지로 Api 공통 응답을 없앤다면 개선이 필요합니다.

kimchijinju reviewed (changes requested):
확인했습니다.
다음부터는 exception handler, 유저 생성 등은 다른 PR로 올라와도 좋을 것 같습니다.

yyysolhhh committed:
chore: 코드 줄바꿈 수정

- 줄바꿈 했을때 http와 너무 멀어서 위로 올림

yyysolhhh committed:
chore: AlreadyExistsException으로 이름 변경

- CustomException을 더 구체적으로 변경

yyysolhhh committed:
feat: ExceptionHandler 응답 처리 방법 수정

GlobalExceptionHandler
- ResponseEntity로 응답 반환

ExceptionResponse
- 예외 응답 객체 정적 팩토리 메서드로 변경

AlreadyExistsException
- CusromException을 구체적인 예외로 이름 변경

yyysolhhh committed:
chore: TODO 주석으로 표시

yyysolhhh committed:
feat: controller에서 ResponseEntity로 응답

UserController
- ResponseEntity로 감싸서 응답하도록 함
- ResponstEntity.created로 생성된 uri 헤더로 반환

Api
- code, status 필드 제거
- 정적 팩토리 메서드로 리팩토링

yyysolhhh committed:
chore: userId를 save메서드에서 반환된 값으로 사용

kimchijinju commented on charlee-web/src/main/java/com/efa73/charleeweb/user/domain/service/UserService.java:
```diff
@@ -0,0 +1,46 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+//@Transactional(readOnly = true)
```
@Transactional(readOnly = true)가 꼭 필요한지? 고민해보기

cyunlee commented on charlee-web/build.gradle:
```diff
@@ -15,7 +15,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-validation'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
-    testImplementation 'org.springframework.security:spring-security-test'
+    testImplementation 'com.h2database:h2'
```
h2의 사용처는 어디인가요?

cyunlee commented on charlee-web/diagram/jwt.md:
보호된 리소스란 정확히 무엇을 의미하나요?

cyunlee commented on charlee-web/src/test/java/com/efa73/charleeweb/config/AdminInitializerTest.java:
```diff
@@ -0,0 +1,62 @@
+    private PasswordEncoder passwordEncoder;
+
+    @Test
+    void 관리자_계정_없으면_생성해야함() {
```
테스트 코드 메서드 이름은 영문으로 짓고 한글 설명은 @DisplayName을 활용하는 것이 어떨까요
